### PR TITLE
Add option to add share link to a file

### LINKED ISSUE #52: Add files from Nextcloud as link [$15]
As we previously talked about, it would be useful to be able to put shared links of files from your Nextcloud directly in your email.

This makes sending mails faster cause attachments don't need to be uploaded. And your files don't leave your server.
When choosing files from your Nextcloud, there could be a split button with the choice of either »Attach« or »Add as shared link«.

<bountysource-plugin>

---
There is a **[$15 open bounty](https://www.bountysource.com/issues/37682663-add-files-from-nextcloud-as-link?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gary-kim**
Resolves #52

This adds the option to share a link to a file from Files.

Also, I was thinking it might be worth including the file sharing service in `@nextcloud/dialogs`.

### CODE REVIEW

File: src/service/FileSharingService.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Makes a share link for a given file or directory.
+ * @param {string} path The file path from the user's root directory
+ * @param {string} token The conversation's token
+ * e.g. `/myfile.txt`
+ */
+const shareFile = async function (path, token) {
+	try {
+		return axios.post(generateOcsUrl('apps/files_sharing/api/v1', 2) + 'shares', {
```
Comment (ChristophWurst) [inline src/service/FileSharingService.js]: don't you miss the `await` here? :thinking:

File: src/service/FileSharingService.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			error.response.data.ocs &&
+			error.response.data.ocs.meta &&
+			error.response.data.ocs.meta.message
+		) {
+			console.error(`Error while sharing file: ${error.response.data.ocs.meta.message || 'Unknown error'}`)
+			showError(error.response.data.ocs.meta.message)
+		} else {
+			console.error(`Error while sharing file: Unknown error`)
+			showError(t('mail', 'Error while sharing file'))
+		}
```
Comment (ChristophWurst) [inline src/service/FileSharingService.js]: we should still (re)throw an error here so the caller knows this wasn't successful

File: src/components/ComposerAttachments.vue [hunk 3]
```diff
@@ -133,12 +136,27 @@ export default {
 				.then((path) => this.emitNewAttachment(this.fileNameToAttachment(path)))
 				.catch((error) => Logger.error('could not choose a file as attachment', {error}))
 		},
+		onAddCloudAttachmentLink() {
+			const picker = getFilePickerBuilder(t('mail', 'Choose a file to share as a link')).build()
+
+			return picker
+				.pick(t('mail', 'Choose a file to share as a link'))
+				.then(async (path) => {
+					const url = (await shareFile(path, getRequestToken())).data.ocs.data.url
```
Comment (ChristophWurst) [inline src/components/ComposerAttachments.vue]: I think the `.data.ocs.data.url` details should go into the service

**Comment by ChristophWurst** [comment]
> Also, I was thinking it might be worth including the file sharing service in `@nextcloud/dialogs`

I think sharing is so complex it would deserve its own package, no? Also this isn't a dialog but just some abstraction of the OCS API

File: src/service/FileSharingService.js (hunk 1 above)
Comment (gary-kim) [reply src/service/FileSharingService.js]: That's from [here](https://github.com/nextcloud/spreed/blob/bc22c21cf70a6243e7df7d838d49018c61468050/src/services/filesSharingServices.js) actually and now that you mention it, yeah, I think it should await there. I'll change it while moving `.data.ocs.data.url`.

**Comment by gary-kim** [comment]
> I think sharing is so complex it would deserve its own package, no?

Your next package when you've got some time? :)

**Comment by gary-kim** [comment]
> Squash and I'll :shipit:

Done!

Also, why is the `:shipit:` emoji a squrriel in a suit and top hat?

**Comment by ChristophWurst** [comment]
> Also, why is the `:shipit:` emoji a squrriel in a suit and top hat?

I don't know, but the internet has answers: https://www.quora.com/On-GitHub-what-is-the-significance-of-the-Ship-It-squirrel :grinning:

**Pull request merged**